File: src/main/java/com/lessmarkup/userinterface/nodehandlers/common/DialogNodeHandler.java
package com.lessmarkup.userinterface.nodehandlers.common;

import com.google.gson.JsonNull;
import com.google.gson.JsonObject;
import com.lessmarkup.Constants;
import com.lessmarkup.TextIds;
import com.lessmarkup.framework.helpers.DependencyResolver;
import com.lessmarkup.framework.helpers.JsonSerializer;
import com.lessmarkup.framework.helpers.LanguageHelper;
import com.lessmarkup.framework.nodehandlers.AbstractNodeHandler;
import com.lessmarkup.interfaces.cache.DataCache;
import com.lessmarkup.interfaces.recordmodel.RecordModelCache;
import com.lessmarkup.interfaces.recordmodel.RecordModelDefinition;
import com.lessmarkup.interfaces.structure.ActionAccess;
import com.lessmarkup.interfaces.structure.Parameter;
import com.lessmarkup.userinterface.model.recordmodel.InputFieldModel;
import com.lessmarkup.userinterface.model.recordmodel.InputFormDefinitionModel;

public abstract class DialogNodeHandler<T> extends AbstractNodeHandler {

    private final InputFormDefinitionModel definitionModel;
    private final DataCache dataCache;
    private final Class<T> modelType;

    protected DialogNodeHandler(DataCache dataCache, Class<T> modelType) {
        this.dataCache = dataCache;
        this.definitionModel = DependencyResolver.resolve(InputFormDefinitionModel.class);
        this.modelType = modelType;
        definitionModel.initialize(modelType);

        for (InputFieldModel field : definitionModel.getFields()) {
            switch (field.getType()) {
                case CODE_TEXT:
                    addScript("scripts/directives/CodeMirrorDirective");
                    break;
                case RICH_TEXT:
                    addScript("scripts/directives/CkEditorDirective");
                    break;
            }
        }
    }

    protected abstract T loadObject();

    protected abstract String saveObject(T changedObject);

    protected String getApplyCaption() {
        return LanguageHelper.getFullTextId(Constants.ModuleType.MAIN, TextIds.APPLY_BUTTON);
    }

    @Override
    public JsonObject getViewData() {
        JsonObject ret = new JsonObject();
        ret.add("definition", definitionModel.toJson());
        Object source = loadObject();
        if (source == null) {
            ret.add("object", JsonNull.INSTANCE);
        } else {
            ret.add("object", JsonSerializer.serializePojoToTree(source));
        }
        ret.addProperty("applyCaption", getApplyCaption());
        return ret;
    }

    @ActionAccess
    public JsonObject save(@Parameter("changedObject") JsonObject changedObject) {
        RecordModelDefinition model = dataCache.get(RecordModelCache.class).getDefinition(modelType);
        model.validateInput(changedObject, false);
        String message = saveObject(JsonSerializer.deserializePojo(modelType, changedObject));
        if (message == null) {
            message = LanguageHelper.getText(Constants.ModuleType.MAIN, TextIds.SUCCESSFULLY_SAVED);
        }
        JsonObject ret = new JsonObject();
        ret.addProperty("message", message);
        return ret;
    }

    @Override
    public String getViewType() {
        return "dialog";
    }
}

File: src/main/scala/com/lessmarkup/userinterface/nodehandlers/common/DialogNodeHandler.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.userinterface.nodehandlers.common

import com.google.gson.{JsonNull, JsonObject}
import com.lessmarkup.framework.helpers.{DependencyResolver, JsonSerializer, LanguageHelper}
import com.lessmarkup.framework.nodehandlers.{AbstractNodeHandler, NodeHandlerConfiguration}
import com.lessmarkup.interfaces.annotations.{ActionAccess, InputFieldType, Parameter}
import com.lessmarkup.interfaces.cache.DataCache
import com.lessmarkup.interfaces.recordmodel.{RecordModelCache, RecordModelDefinition}
import com.lessmarkup.userinterface.model.recordmodel.InputFormDefinitionModel
import com.lessmarkup.{Constants, TextIds}

abstract class DialogNodeHandler[T <: AnyRef](
  dataCache: DataCache,
  modelType: Class[T],
  configuration: NodeHandlerConfiguration
) extends AbstractNodeHandler(configuration) {

  private final val definitionModel: InputFormDefinitionModel = DependencyResolver.resolve(classOf[InputFormDefinitionModel], modelType)

  override def getScripts: Seq[String] = {
    super.getScripts ++ definitionModel.getFields.flatMap(f => {
      f.getType match {
        case InputFieldType.CODE_TEXT => Option("scripts/directives/CodeMirrorDirective")
        case InputFieldType.RICH_TEXT => Option("scripts/directives/CkEditorDirective")
        case _ => None
      }
    })
  }

  protected def loadObject: Option[T]

  protected def saveObject(changedObject: Option[T]): String

  protected def getApplyCaption: String = {
    LanguageHelper.getFullTextId(Constants.ModuleTypeMain, TextIds.APPLY_BUTTON)
  }

  override def getViewData: Option[JsonObject] = {
    val ret: JsonObject = new JsonObject
    ret.add("definition", definitionModel.toJson)
    val source = loadObject
    if (source.isEmpty) {
      ret.add("object", JsonNull.INSTANCE)
    }
    else {
      ret.add("object", JsonSerializer.serializePojoToTree(source.get))
    }
    ret.addProperty("applyCaption", getApplyCaption)
    Option(ret)
  }

  @ActionAccess
  def save(@Parameter("changedObject") changedObject: JsonObject): JsonObject = {
    val model: RecordModelDefinition = dataCache.get(classOf[RecordModelCache]).getDefinition(modelType).get
    model.validateInput(changedObject, isNew = false)
    var message: String = saveObject(JsonSerializer.deserializePojo(modelType, changedObject))
    if (message == null) {
      message = LanguageHelper.getText(Constants.ModuleTypeMain, TextIds.SUCCESSFULLY_SAVED)
    }
    val ret: JsonObject = new JsonObject
    ret.addProperty("message", message)
    ret
  }

  override def getViewType = "dialog"
}
